subsys: testsuite: make tc_util overridable

torsteingrindvik opened the pull request:
**What:**
A Kconfig variable (under menuconfig->Testing) is added.
Checking this prompts subsys/testsuite/include/tc_util.h to look for tc_util_user_override.h.
This user override include can define various things in tc_util.h to control how output is produced during execution of test suites.

If not checked, no changes are made to current projects.

**Why:**
Some test suites with many tests (hundreds) are overly verbose---this feature allows projects to tune the output as necessary.

SebastianBoe commented:
Generally, if something is not supported, e.g. configurability of verbosity, it is better to add this support than to swap the implementation out with something that does support it.

Could we look into extending tc_util to support configurable verbosity?

torsteingrindvik commented:
> 
> 
> Generally, if something is not supported, e.g. configurability of verbosity, it is better to add this support than to swap the implementation out with something that does support it.
> 
> Could we look into extending tc_util to support configurable verbosity?

Yes that would be useful.
However the intent of this was slightly different, as this allows the user full customization over how results are displayed whereas verbosity levels normally only allows a set of pre-defined choices.

nashif commented:
Can you please give examples of tests that is overly verbose and why this actually matters?

torsteingrindvik commented:
> 
> 
> Can you please give examples of tests that is overly verbose and why this actually matters?

We are working on crypto tests that have several hundreds of test cases (stemming from [here](https://www.nist.gov/programs-projects/cryptographic-algorithm-validation-program) mostly). Multiply this with testing against several backends results in a very large amount of output.
A reduction of ~4 lines per test case to ~1 reduces visual noise, which is how we would use this functionality.

It's a quality of life improvement, I'm not sure what qualifies as mattering or not.

nashif commented on subsys/testsuite/include/tc_util.h:
```diff
@@ -15,7 +15,13 @@
 #include <sys/printk.h>
 
+#if defined CONFIG_ZTEST_TC_UTIL_USER_OVERRIDE
+#include <tc_util_user_override.h>
```
Is this going to be included per test wishing to override? Can you please add a test that uses this new functionalty? We also need documentation for this in doc/guides/test/

torsteingrindvik replied:
I added a test and some documentation in the latest commit.

nashif commented:
> It's a quality of life improvement, I'm not sure what qualifies as mattering or not.

Doesn't the verbosity depend on how you actually write your test? So why would you provide a verbose test in the first place if you do not need the verbose message?
Also, is the output supposed to be observed by humans? I most cases this would automated and scripts have no quality of life. 
I guess it is difficult to understand the intentions without seeing the test.

Btw, is this test going to be submitted to the zephyr tree?

nvlsianpu commented on tests/ztest/custom_output/include/tc_util_user_override.h:
```diff
@@ -0,0 +1,31 @@
+#ifndef __TC_UTIL_USER_OVERRIDE_H__
```
missing license

torsteingrindvik replied:
Added.

nvlsianpu commented on tests/ztest/custom_output/prj.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_ZTEST=y
```
new line at the end of file

torsteingrindvik replied:
Added.

torsteingrindvik commented:
> Doesn't the verbosity depend on how you actually write your test? So why would you provide a verbose test in the first place if you do not need the verbose message?
> Also, is the output supposed to be observed by humans? I most cases this would automated and scripts have no quality of life.
> I guess it is difficult to understand the intentions without seeing the test.
> 
> Btw, is this test going to be submitted to the zephyr tree?

It is the output produced by ztest itself that is changed, not what the test itself produces.
It should make sense if you run the test I added (the output is supressed though, I needed -vv to see it via sanitycheck).

The test application I will be using this PR for will be merged into NordicPlayground/fw-nrfconnect-nrf first, but might be submittet here at some later point.

nvlsianpu commented on tests/ztest/custom_output/prj_customized_output.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_ZTEST=y
+CONFIG_ZTEST_TC_UTIL_USER_OVERRIDE=y
```
new line at the end of file

torsteingrindvik replied:
Added.

nvlsianpu commented on tests/ztest/custom_output/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources})
+zephyr_include_directories(include)
```
new line at the end of file

torsteingrindvik replied:
Added.

nvlsianpu commented on tests/ztest/custom_output/src/main.c:
```diff
@@ -0,0 +1,39 @@
+
+static void test_assert_fail(void)
+{
+	ztest_test_fail();
```
if this test failing it cause whole shippable to failing - this can be something rather for sample

torsteingrindvik replied:
I removed this and kept only the passing and skipped tests---that is enough to show the new output anyway.

nvlsianpu commented on doc/guides/test/ztest.rst:
```diff
@@ -341,3 +341,22 @@ expect the values ``a=2`` and ``b=3``, and telling ``returns_int`` to return
    :project: Zephyr
+
+Customizing Test Output
+**************************
```
align asterisks line to line above

torsteingrindvik replied:
Fixed.

nvlsianpu commented:
Doc error 

>/var/lib/shippable/build/IN/main_repo/zephyr/doc/_build/rst/doc/guides/test/ztest.rst:361: WARNING: 'any' reference target not found: zephyr_include_directories(my_folder)

Doxygen is trying to find a reference symbol in parsed code (as it would be C function reference).

karstenkoenig commented:
> Doc error
> 
> > /var/lib/shippable/build/IN/main_repo/zephyr/doc/_build/rst/doc/guides/test/ztest.rst:361: WARNING: 'any' reference target not found: zephyr_include_directories(my_folder)
> 
> Doxygen is trying to find a reference symbol in parsed code (as it would be C function reference).
> 
> @karstenkoenig need to change this to non-reference. @dbkinder, @b-gent Can you help with this?

Did you mean to ping me? @torsteingrindvik wrote that line, I don't think I can help

torsteingrindvik commented:
@nvlsianpu I don't think the doc is an issue at this point.
The Shippable check is complaining, and I am not sure how to resolve it.

dbkinder commented on doc/guides/test/ztest.rst:
```diff
@@ -341,3 +341,22 @@ expect the values ``a=2`` and ``b=3``, and telling ``returns_int`` to return
+        #define SOMETHING <default implementation>
+        #endif /* SOMETHING */
+
+The line :func:`zephyr_include_directories(my_folder)` should be added to
```
Fix the doc build error with:
```
Add the line ``zephyr_include_directories(my_folder)`` to
your project's :file:`CMakeLists.txt` to let Zephyr find your header file during builds.
```

torsteingrindvik replied:
Changed it to your suggestion.

dbkinder commented on doc/guides/test/ztest.rst:
```diff
@@ -341,3 +341,22 @@ expect the values ``a=2`` and ``b=3``, and telling ``returns_int`` to return
+
+Customization is enabled by setting :option:`CONFIG_ZTEST_TC_UTIL_USER_OVERRIDE` to "y".
+
+Then, a file :file:`tc_util_user_override.h` is expected to be found by Zephyr.
```
Combine these two sentences for clarity:
```
Then, create a file named :file:`tc_util_user_override.h` with your customizations.
```

torsteingrindvik replied:
I reordered the paragraph slightly to hopefully increase clarity.

torsteingrindvik committed:
subsys/testsuite: make tc_util overridable

A Kconfig boolean is added to allow users to provide their own
output strings when running tests via ztest.
This allows changing e.g. the PASS/FAIL/SKIPPED strings,
add counters, change separators, and similar.

A test using the feature and relevant documentation is added.

Signed-off-by: Torstein Grindvik <torstein.grindvik@nordicsemi.no>

dbkinder reviewed (approved):
docs LGTM, thanks!
